pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java

```diff
@@ -272,17 +288,39 @@ private PersistentSubscription createPersistentSubscription(String subscriptionN
 
     @Override
     public void publishMessage(ByteBuf headersAndPayload, PublishContext publishContext) {
-        MessageDeduplication.MessageDupStatus status = messageDeduplication.isDuplicate(publishContext, headersAndPayload);
-        switch (status){
-            case NotDup:
-                ledger.asyncAddEntry(headersAndPayload, this, publishContext);
-                break;
-            case Dup:
-                // Immediately acknowledge duplicated message
-                publishContext.completed(null, -1, -1);
-                break;
-            default:
-                publishContext.completed(new MessageDeduplication.MessageDupUnknownException(), -1, -1);
+        pendingWriteOps.incrementAndGet();
+        if (isFenced) {
+            publishContext.completed(new TopicFencedException("fenced"), -1, -1);
+            decrementPendingWriteOpsAndCheck();
+            return;
+        } else {
+            MessageDeduplication.MessageDupStatus status = messageDeduplication.isDuplicate(publishContext, headersAndPayload);
+            switch (status) {
+                case NotDup:
+                    ledger.asyncAddEntry(headersAndPayload, this, publishContext);
+                    break;
+                case Dup:
+                    // Immediately acknowledge duplicated message
+                    publishContext.completed(null, -1, -1);
+                    pendingWriteOps.decrementAndGet();
```

Comment:
Each time we decrement we need to check whether we need to notify. We should just keep calling `decrementPendingWriteOpsAndCheck()`

Reply:
will change